component/common/loki/client/internal/marker_file_handler.go

```diff
@@ -0,0 +1,107 @@
+package internal
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/agent/component/common/loki/wal"
+	"github.com/natefinch/atomic"
+)
+
+const (
+	MarkerFolderName = "remote"
+	MarkerFileName   = "segment_marker"
+
+	MarkerFolderMode os.FileMode = 0o700
+	MarkerFileMode   os.FileMode = 0o600
+)
+
+// MarkerFileHandler is a file-backed wal.Marker, that also allows one to write to the backing store as particular
+// segment number as the last one marked.
+type MarkerFileHandler interface {
+	wal.Marker
+
+	// MarkSegment writes in the backing file-store that a particular segment is the last one marked.
+	MarkSegment(segment int)
+}
+
+type markerFileHandler struct {
+	logger                    log.Logger
+	lastMarkedSegmentDir      string
+	lastMarkedSegmentFilePath string
+}
+
+var (
+	_ MarkerFileHandler = (*markerFileHandler)(nil)
+)
+
+// NewMarkerFileHandler creates a new markerFileHandler.
+func NewMarkerFileHandler(logger log.Logger, walDir string) (MarkerFileHandler, error) {
+	markerDir := filepath.Join(walDir, MarkerFolderName)
+	// attempt to create dir if doesn't exist
+	if err := os.MkdirAll(markerDir, MarkerFolderMode); err != nil {
+		return nil, fmt.Errorf("error creating segment marker folder %q: %w", markerDir, err)
+	}
+	// do we need this bit below?
+	//if stat, err := os.Stat(markerDir); err == nil && stat.Mode().Perm() != MarkerFileMode {
+	//	if err := os.Chmod(markerDir, MarkerFileMode); err != nil {
+	//		return nil, err
+	//	}
+	//}
+
```

Comment:
@mattdurham do you think we need this bit? Like if the folder exists, double check it has the perms we want, and if not, do a chmod?

Reply:
No, since we are within the agents datapath we can assume we own it.

Reply:
Just added the logic, so will remove it!